feat(NRQL tutorial): NRQL introductory tutorial (71564) (#10794)
newrelic/docs-website: src/content/docs/query-your-data/nrql-new-relic-query-language/get-started/introduction-nrql-tutorial.mdx

---
title: "Introductory NRQL tutorial: unlock your data"
tags:
  - NRQL tutorial
  - 'NRQL: New Relic query language'
  - Get started
  - Tutorial
translate:
  - jp
  - kr
metaDescription: "Learn how to use NRQL with this basic, introductory tutorial."
redirects:
---

> We've covered the basics of NRQL in our other topics, such as what NRQL is, how NRQL works, and how NRQL relates to charts and dashboards. With this information, beginning to work with NRQL is the next step, and this tutorial is designed to help you do just that! Below we'll cover some foundational concepts within the language and get you set up with creating a some basic NRQL queries using your data. This tutorial will cover:

Author: - I wanted to have a sense of progression through the docs using the previous topics we've created/revamped and call out that readers may want to read them before starting the tutorial, but I'm unsure about the wording fitting with our language conventions. Any thoughts would be appreciated!

Reviewer: you might have fixed it, didn't check, but typo of "creating a some basic..."

* Use `SELECT` and `FROM`
* Use `LIMIT`
* Specify attributes
* Aggregate using `average()`, `max()`, `min()`, `sum()`, and `count()`
* Use `SINCE` and `UNTIL`
* Plot line graphs with `TIMESERIES`
* Filter using `WHERE`
* Group attributes using `FACET`

Want to follow this tutorial within the New Relic platform? We've included a NRQL Lessons application on our [opensource.newrelic.com](https://opensource.newrelic.com/projects/newrelic/nr1-learn-nrql) with the same steps and examples to help you quickly see NRQL's value and power within the application. The video below shows you how to install the NRQL Lesson.

<Video
  id="3UHEHbprs6I"
  type="youtube"
/>

Ready to get started? Let's go!

<CollapserGroup>
  <Collapser
    className="freq-link"
    id="first-query"
    title="1. Your first query"
  >
  <SideBySide>
  <Side>

> Let's start by looking at a single event type in NRDB called **Transaction**, gathered by New Relic APM. Every NRQL query must have `SELECT` and `FROM` clauses: you must `SELECT` some data and tell us where it's `FROM`. Start with this basic query that selects everything from the **Transaction** event type.

Reviewer: I'd recommend using code format instead of bold, so `Transaction` instead of **Transaction**. 

Instead of 'selects everything', as that may be ambiguous as to what that means, maybe 'returns the`Transaction` events reported in your New Relic account' or something.

Author: Roger, I'll get this change made. Thank you!

    LIMIT 1
    ```
  </Side>
  <Side>
 <img
  title="Select, From, and Limit query"
  alt="A screenshot of a query with Select, From, and Limit"
  src={queriesnrqltutorial2}
/>
  </Side>
</SideBySide>
<SideBySide>
  <Side>
    We're now controlling the volume of the results. But what if we don't want all the attributes? What if we would prefer to see only specific data points? Like SQL, we can do this with only a few characters by replacing `*` with the name of the attribute(s) we want. In this case, we'll ask for the name of a transaction and the duration of time it took.

    ```sql
    SELECT name, duration
    FROM Transaction
    ```
    
  </Side>
  <Side>
  <img
  title="Select and From with Duration query"
  alt="A screenshot of a query with Select and From using Duration"
  src={queriesnrqltutorial3}
/>
  </Side>
</SideBySide>
With this, you've completed queries that can return either all or some specific attributes of an event, as well as querying a specific number of events. Let's go to the next step: aggregating data.
  </Collapser>
  <Collapser
    className="freq-link"
    id="aggregate-data"
    title="2. Aggregate data"
  >
<SideBySide>
  <Side>

> Sometimes you may want to view summarized aggregate data across many events instead of viewing individual events. This is where NRDB shines. NRDB can scan billions of events and provide instant, real-time answers to questions about your data. For example, each `Transaction` event has a duration attribute that represents how many seconds the call took to execute. Let's see if we can discover the average duration across all `Transaction` events:

Reviewer: instead of 'event', I think we should use 'data points'. The reason being: some of the data types we can query with NRQL include `Metric` and `Span` and `Log` data, and we avoid calling those 'events' (trying to keep those MELT data types separate). So even though we often internally refer to everything in NRDB as 'events', we should try as much as we can to avoid comingling those terms. So instead I think it's just as easy to say 'NRDB can scan billions of data points...' and such.

Author: Sounds good to me! Thank you!

  src={queriesnrqltutorial4}
/>
  </Side>
</SideBySide>
<SideBySide>
  <Side>
NRQL has many built-in functions you can use to aggregate event data. Commonly used functions include `max()`, `min()`,`average()`, `median()`, and `percentile()`. In the following query, we'll ask NRDB for the duration of the slowest transaction from the last 60 minutes To do this, we will ask NRDB for the `max(duration).`
```sql
SELECT max(duration) 
FROM Transaction
```
  </Side>
  <Side>
    <img
  title="Select maximum duration from transaction query"
  alt="A screenshot of a query selecting the maximum duration from transaction"
  src={queriesnrqltutorial5}
/>
  </Side>
</SideBySide>
<SideBySide>
  <Side>
We now know how long the slowest customer experience in the last 60 minutes took. Next, let's find the inverse and see how long the fastest experience was by asking NRDB for the `min(duration)`.
```sql
SELECT min(duration) 
FROM Transaction
```
  </Side>
  <Side>
    <img
  title="Select minimum duration from transaction query"
  alt="A screenshot of a query selecting the minimum duration from transaction"
  src={queriesnrqltutorial6}
/>
  </Side>
</SideBySide>
<SideBySide>
  <Side>

> As you can see, `max()` and `min()` both return important data points. However, we may also want to perform arithmetic on our data. With 'sum()', we can do basic addition on a numeric attribute. In this query, we'll use a new attribute, `databaseCallCount`. This attribute counts the number of database calls each transaction makes. By adding them, we can get the total volume of database calls in the default 60 minute time window (this will be 0 if your apps don't talk to a database).

Reviewer: ```suggestion
As you can see, `max()` and `min()` both return important data points. However, we may also want to perform arithmetic on our data. With `sum()`, we can do basic addition on a numeric attribute. In this query, we'll use a new attribute, `databaseCallCount`. This attribute counts the number of database calls each transaction makes. By adding them, we can get the total volume of database calls in the default 60 minute time window (this will be 0 if your apps don't talk to a database).
```

Author: Thank you!

  </Collapser>
  <Collapser
    className="freq-link"
    id="time-ranges"
    title="3. Use time ranges"
  >
<SideBySide>
  <Side>
    Every piece of data in NRDB has a `timestamp`, and every query operates on a subset of data within a time range. If you don't provide a time range in your NRQL query, it will return the last 60 minutes by default as we saw in the previous lesson. But what if we want to see data from a different time range? How do we control the window of data we see? The answer to these questions is very straightforward: use `SINCE` and `UNTIL` keywords to provide a beginning and end to your query's time range.
```sql 
SELECT average(duration) 
FROM Transaction 
SINCE 1 day ago
```
  </Side>
  <Side>
    <img
  title="Select the average duration since 1 day ago query"
  alt="A screenshot of a query selecting the average duration since 1 day ago"
  src={queriesnrqltutorial9}
/>
  </Side>
</SideBySide>
<SideBySide>
  <Side>

>     We can also specify time ranges relative to when we make our queries by using keywords like `day`, `week`, `hour`, `minute`, or their plural equivalents. We can also use logical expressions like `SINCE` today or `SINCE` this week.

Reviewer: ```suggestion
    We can also specify time ranges relative to when we make our queries by using keywords like `day`, `week`, `hour`, `minute`, or their plural equivalents. We can also use logical expressions like `SINCE today` or `SINCE this week`.
```

Author: Thank you!

```sql
SELECT average(duration) 
FROM Transaction 
SINCE this week
```
  </Side>
  <Side>
    <img
  title="Select the average duration since 1 week ago"
  alt="A screenshot of a query selecting the average duration since 1 week ago"
  src={queriesnrqltutorial10}
/>
  </Side>
</SideBySide>
<SideBySide>
  <Side>

>     Here is another query that demonstrate relative time ranges. Note that when specifying a relative time, you need to include the keyword `ago`.

Reviewer: ```suggestion
    Here is another query that demonstrate relative time ranges. Note that when specifying a relative time, you need to include the keyword `ago`.
```

Author: I'm not sure I'm seeing where the change is here? It's showing a difference but I can't see where that difference is.

  </Collapser>
  <Collapser
    className="freq-link"
    id="time-series"
    title="4. Query time series"
  >
<SideBySide>
  <Side>
    A common NRQL use case is querying how values change over time. This type of query provides data for line charts, area charts, and other visualizations. We've already run some queries to explore aggregation functions `average()`, `max()`, `min()`, `sum()`, and `count()`. Now, we can take any query made with aggregation functions and add the `TIMESERIES` keyword to plot the values over time.
```sql
SELECT average(duration) 
FROM Transaction 
SINCE 1 day ago 
TIMESERIES
```
  </Side>
  <Side>
    <img
  title="Select the average duration since 1 day ago over with a timeseries"
  alt="A screenshot of a query that selects the average duration since 1 day ago with a timeseries"
  src={queriesnrqltutorial12}
/>
  </Side>
</SideBySide>
<SideBySide>

>   <Side>

Reviewer: probably can just do "A timeseries query..."

Author: Thank you!

    ```sql
    SELECT average(duration) 
    FROM Transaction 
    SINCE 1 day ago 
    TIMESERIES MAX
    ```
  </Side>
  <Side>
    <img
  title="Select the average duration since 1 day ago using the maximum number of buckets"
  alt="A screenshot of a query selecting the average duration since 1 ago using the maximum number of buckets"
  src={queriesnrqltutorial14}
/>
  </Side>
</SideBySide>
Now we're achieving new levels of visualization, breaking out from summarized numbers to line charts of data trends over time. As you can see, using `TIMESERIES` offers full control over visualizations, granularity, and averaging data over specified windows. Let's go to the next step: using the 'Where' clause.
  </Collapser>
  <Collapser
    className="freq-link"
    id="where"
    title="5. Use Where clauses"
  >
<SideBySide>
  <Side>

>     You may find that sometimes, you want your query to operate on a specific subset of NRDB data. For example, let's query our application's average response time. This data includes an attribute called `transactionType`, which specifies whether the transaction we timed was a "Web" or "Non-Web" (such as a cron job or queued task) transaction. Let's filter our query to only return data for the "Web" transaction type:

Reviewer: i think it's okay to use code format for these things instead of quotes, so like 
was a `web` or `non-web` transaction
Think it's easier to parse

    ```sql
    SELECT average(duration) 
    FROM Transaction 
    WHERE transactionType='Web' 
    TIMESERIES
    ```
  </Side>
  <Side>
    <img
  title="Select the avarege duration for Web transaction types with a timeseries"
  alt="A screenshot of a query selecting the average duration for Web transaction types with a timeseries"
  src={queriesnrqltutorial15}
/>
  </Side>
</SideBySide>
<SideBySide>
  <Side>

>     You can also combine `AND` and `OR` to make more complex queries, using parentheses "(" and ")" to control how these work together. This query shows the average response time of "Web" transactions that were both faster than 100ms, and had a response code of either 200 or 302. Notice how we use parentheses to control the `OR`.

Reviewer: think you can use code format here, too, so like
using parentheses `(` and `)`...

Author: Good call! Thank you!

    ```sql
    SELECT average(duration) 
    FROM Transaction 
    FACET name 
    SINCE 3 hours ago 
    LIMIT 20
    ```
  </Side>
  <Side>
    <img
  title="Select the average duration from the top 20 results since 3 hours ago grouped by name without a timeseries"
  alt="A screenshot of a query selecting the average duration from the top 20 results since 3 hours ago grouped by name without a timeseries"
  src={queriesnrqltutorial19}
/>
  </Side>
</SideBySide>
<SideBySide>
  <Side>
  Finally, here's a slightly more complex query that compares the quantity of Web transactions, broken down by individual applications that report to New Relic:
  ```sql
  SELECT count(*) 
  FROM Transaction 
  WHERE transactionType='Web' 
  FACET appName 
  LIMIT 5 
  SINCE 6 hours ago 
  TIMESERIES
  ```
  </Side>
  <Side>
    <img
  title="Select all web transactions from the top 5 results since 6 hours ago grouped by appName with a timeseries"
  alt="A screenshot of a query selecting all web transactions from the top 5 results since 6 hours ago grouped by appName with a timeseries"
  src={queriesnrqltutorial20}
/>
  </Side>
</SideBySide>

> Grouping or "faceting" is important: it allows us to get dimensional with our data. Using the `FACET` keyword is the simplest way to do this. When applying an aggregation function, we can use `FACET` to group by any attribute, default or custom, that reports event data (just like the `WHERE` clause). We recommend tagging on custom data to your events. This allows you to more easily investigate the resulting data set.

Reviewer: the 'tagging on' is a bit ambiguous and potentially confusing, as it overlaps with 'tags'. I'd say maybe just 'We recommend adding custom attributes to your data, or reporting new custom events...' and maybe linking to those docs.

Author: Good idea, thank you!

---

> title: "Introductory NRQL tutorial: unlock your data"

Author: Hmmm, good callout, I will look into this!

tags:
  - NRQL tutorial
  - 'NRQL: New Relic query language'
  - Get started
  - Tutorial

> translate:

Author: I'm not sure, hopefully someone else can confirm? If not I will find out before we publish, thanks for calling this out!

> translate:

Reviewer 2: @jeff-colucci You don't need to add the translate part of the frontmatter.

> translate:

Author: Thank you! I'll get it taken out

  - jp
  - kr

> metaDescription: "Learn how to use NRQL with this basic, introductory tutorial."

Author: Good call, I will drop basic as we are selling this as an introductory tutorial elsewhere.

> redirects:

Reviewer 2: You can also leave redirects out if there are no redirects

Author: Thank you!

import queriesnrqltutorial8 from 'images/queries-nrql_screenshot-crop-tutorial-8.png'

import queriesnrqltutorial9 from 'images/queries-nrql_screenshot-crop-tutorial-9.png'

import queriesnrqltutorial10 from 'images/queries-nrql_screenshot-crop-tutorial-10.png'

import queriesnrqltutorial11 from 'images/queries-nrql_screenshot-crop-tutorial-11.png'

import queriesnrqltutorial12 from 'images/queries-nrql_screenshot-crop-tutorial-12.png'

import queriesnrqltutorial13 from 'images/queries-nrql_screenshot-crop-tutorial-13.png'

import queriesnrqltutorial14 from 'images/queries-nrql_screenshot-crop-tutorial-14.png'

import queriesnrqltutorial15 from 'images/queries-nrql_screenshot-crop-tutorial-15.png'

import queriesnrqltutorial16 from 'images/queries-nrql_screenshot-crop-tutorial-16.png'

import queriesnrqltutorial17 from 'images/queries-nrql_screenshot-crop-tutorial-17.png'

import queriesnrqltutorial18 from 'images/queries-nrql_screenshot-crop-tutorial-18.png'

import queriesnrqltutorial19 from 'images/queries-nrql_screenshot-crop-tutorial-19.png'

import queriesnrqltutorial20 from 'images/queries-nrql_screenshot-crop-tutorial-20.png'

> 

Author: I do have some guideposts in the works at [this PR](https://github.com/newrelic/docs-website/pull/10831). Do you think this one works or should we move it/add something else that's more visible? I do plan on going investigating all NRQL topics and putting in links to tutorial in the appropriate ones as well.

> 

Author: Maybe we can eliminate that as a callout and integrate it into the "guidepost" that was added?

> We've covered the basics of NRQL in our other topics, such as [what NRQL is](/docs/query-your-data/nrql-new-relic-query-language/get-started/introduction-nrql-new-relics-query-language/), [how NRQL works](docs/query-your-data/nrql-new-relic-query-language/get-started/introduction-nrql-how-nrql-works/), and [how NRQL relates to charts and dashboards](docs/query-your-data/nrql-new-relic-query-language/get-started/charts-and-dashboards-with-nrql/). With this information, beginning to work with NRQL is the next step, and this tutorial is designed to help you do just that! Below we'll cover some foundational concepts within the language and get you set up with creating a some basic NRQL queries using your data. This tutorial teaches you how to:

Author: Was missing the slash on the link, it should be in now. I like your suggestions for the sentences and have made changes. Thank you!

* Use `SELECT` and `FROM`
* Use `LIMIT`
* Specify attributes
* Aggregate using `average()`, `max()`, `min()`, `sum()`, and `count()`
* Use `SINCE` and `UNTIL`
* Plot line graphs with `TIMESERIES`
* Filter using `WHERE`
* Group attributes using `FACET`

> Want to learn NRQL within the New Relic platform? We've included a NRQL Lessons application on our [opensource.newrelic.com](https://opensource.newrelic.com/projects/newrelic/nr1-learn-nrql) site to help you quickly see NRQL's value and power within the application. The video below shows you how to install the NRQL Lesson.

Author: Agree I didn't do as a good a job with this. This tutorial is basically the same thing as the NRL app, I just changed a lot of the styling to better suit our online docs to increase visibility on this information. So really the NRL lesson is an in-app alternative with the same information. I've made changes, let me know if you think it gets the point across better!

<Video
  id="3UHEHbprs6I"
  type="youtube"
/>

Ready to get started? Let's go!

<CollapserGroup>
  <Collapser
    className="freq-link"
    id="first-query"
    title="1. Your First Query"
  >
  <SideBySide>
  <Side>
Let's start by looking at a single event type in NRDB called **Transaction**, gathered by New Relic APM. Every NRQL query must have `SELECT` and `FROM` clauses: you must `SELECT` some data and tell us where it is `FROM`. Start with this basic query that selects everything from the **Transaction** event type.
 
    ```sql
    SELECT *
    FROM Transaction
    ```

  </Side>
  <Side>
    <img
  title="Select and From query"

>   alt="A screenshot of a query with Select and From"

Author: I'm thinking about adding the titles as fig captions. We do have the room for them on the right side, but they are rendered quite small and I was worried about potential text bloat. What do you think?

  src={queriesnrqltutorial1}
/>
  </Side>
  </SideBySide>
  <SideBySide>
  <Side>
 This returns a lot of results, each with a timestamp and a collection of attributes. For now, we only want a single result, so let's limit the results to a single record using `LIMIT 1`.

>  <Callout variant="important"> When a LIMIT isn't supplied the default will be used, which is 100 table rows for SELECT * queries or 10 aggregated values for FACET queries and SELECT (attributes) queries. You can specify any limit up to the maximum. Use LIMIT MAX to return the maximum number of results possible.

Author: A bit stumped on this callout, I feel like it looks kind of weird here (it looks worse after the query where I originally had it), but it seems important enough to know this information that some sort of callout should be made. Any suggestions are more than welcome.

Reviewer: might consider using some code formatting for some of this, like 

`SELECT *` queries
`SELECT (attributes)` queries

Author: Thank you!

</Callout>

    ```sql
    SELECT *
    FROM Transaction
    LIMIT 1
    ```
  </Side>
  <Side>
 <img
  title="Select, From, and Limit query"
  alt="A screenshot of a query with Select, From, and Limit"
  src={queriesnrqltutorial2}
/>
  </Side>
</SideBySide>
<SideBySide>
  <Side>

>     We're now controlling the volume of the results. But what if we don't want all the attributes? What if we would prefer to see only specific data points? Like SQL, we can do this with only a few characters by replacing * with the name of the attribute(s) we want. In this case, we'll ask for the name of a transaction and the duration of time it took.

Author: Agreed, thank you!

    ```sql
    SELECT name, duration
    FROM Transaction
    ```
    
  </Side>
  <Side>
  <img
  title="Select and From with Duration query"
  alt="A screenshot of a query with Select and From using Duration"
  src={queriesnrqltutorial3}
/>
  </Side>
</SideBySide>
With this, you've completed queries that can return either all or some specific attributes of an event, as well as querying a specific number of events. Let's go to the next step: aggregating data.

>   </Collapser>

Author: I wanted to have a better separation too! I was thinking of asking Eng to see if something can be done about that as well as a few other things after everyone is back from the holidays.

>   </Collapser>

Reviewer 2: @jeff-colucci Please do ask the eng team if they can help with this. I also really like this side-by-side approach. I don't know if this is a better option, but I kind of wanted to see the screenshots next to the codeblocks rather than the explanation.

OR I'd want to see the screenshot on the left and the explanation on the right. This might help break things up a little bit given the current limitations?

>   </Collapser>

Author: I like the idea of swapping the images and text. I thought that the query examples before the explanations were a little wonky (which would fix the positioning issue), but maybe there is a way or there can be a way to center or bottom justify the images in the side-by-sides which would match the positioning.

  <Collapser
    className="freq-link"
    id="aggregate-data"

>     title="2. Aggregate Data"

Author: Thank you, don't know why I did half correctly and half incorrectly!

  >
<SideBySide>
  <Side>
Sometimes you may want to view summarized aggregate data across many events instead of viewing individual events. This is where NRDB shines. NRDB can scan billions of events and provide instant, real-time answers to questions about your data. For example, each `Transaction` event has a duration attribute that represents how many seconds the call took to execute. Let's see if we can discover the average duration across all `Transaction` events:
```sql
SELECT average(duration) 
FROM Transaction
```
  </Side>
  <Side>
    <img
  title="Select average duration from transaction query"
  alt="A screenshot of a query selecting the average duration from transaction"
  src={queriesnrqltutorial4}
/>
  </Side>
</SideBySide>
<SideBySide>
  <Side>

> NRQL has many built-in functions you can use to aggregate event data. Common functions you might use include `max()`, `min()`, `median()`, and `percentile()`. In the following query, we'll ask NRDB for the duration of the slowest transaction from the last 60 minutes To do this, we will ask NRDB for the `max(duration).`

Author: Good call, thanks!

```sql
SELECT max(duration) 
FROM Transaction
```
  </Side>
  <Side>
    <img
  title="Select maximum duration from transaction query"
  alt="A screenshot of a query selecting the maximum duration from transaction"
  src={queriesnrqltutorial5}
/>
  </Side>
</SideBySide>
<SideBySide>
  <Side>

> We now know how long the slowest customer experience in the last 60 minutes took. Next, let's find the inverse and see how long the fastest experience was by asking NRDB for the `min(duration)`.

Author: Thank you, I will be taking a look at this voice and consolidating it tomorrow. I will take some time to chew on if I want it to be "you" or "we", maybe after we get some feedback on what works better (I'm leaning "we" right now)

> We now know how long the slowest customer experience in the last 60 minutes took. Next, let's find the inverse and see how long the fastest experience was by asking NRDB for the `min(duration)`.

Reviewer 2: In general, we prefer second person (you/your). The we is pretty far afield from what we (haha) usually do. I'm still mulling it over, but I'm going to have a tough time breaking away from my impulse toward the second person.

> We now know how long the slowest customer experience in the last 60 minutes took. Next, let's find the inverse and see how long the fastest experience was by asking NRDB for the `min(duration)`.

Author: I will certainly go through and smooth out the voice early next week. 2 votes for "you" is more than enough votes for me!

  alt="A screenshot of a query selecting the average duration since 1 day ago"
  src={queriesnrqltutorial9}
/>
  </Side>
</SideBySide>
<SideBySide>
  <Side>
    We can also specify time ranges relative to when we make our queries by using keywords like `day`, `week`, `hour`, `minute`, or their plural equivalents. We can also use logical expressions like `SINCE` today or `SINCE` this week.
```sql
SELECT average(duration) 
FROM Transaction 
SINCE this week
```
  </Side>
  <Side>
    <img
  title="Select the average duration since 1 week ago"
  alt="A screenshot of a query selecting the average duration since 1 week ago"
  src={queriesnrqltutorial10}
/>
  </Side>
</SideBySide>
<SideBySide>
  <Side>
    Here is another query that demonstrate relative time ranges. Note that when specifying a relative time, you need to include the keyword `ago`.
```sql
SELECT average(duration) 
FROM Transaction 
SINCE 1 week ago 
UNTIL 2 days ago
```
  </Side>
  <Side>
    <img
  title="Select the average duration since 1 week ago until 2 days ago"
  alt="A screenshot of a query selecting the average duration since 1 week ago until 2 days ago"
  src={queriesnrqltutorial11}
/>
  </Side>
</SideBySide>

> Now we're getting even more control of our data. By adding a time frame to a query, we're define exactly which window of data we want to see. These controls let us zero in on the data we want by defining where the data reporting should start and end. In our examples, we saw that we can easily choose the window and use common terms like `week`, `day`, `hour`, `minute`, and their plurals. You can even use `month` and `year` if you're storing that much data in New Relic. Let's go to the next step: querying a time series.

Reviewer 3: @jeff-colucci looks like a typo. Should be "...we're defining exactly...."

Author: Thank you!

> Now we're getting even more control of our data. By adding a time frame to a query, we're defining exactly which window of data we want to see. These controls let us zero in on the data we want by defining where the data reporting should start and end. In our examples, we saw that we can easily choose the window and use common terms like `week`, `day`, `hour`, `minute`, and their plurals. You can even use `month` and `year` if you're storing that much data in New Relic. Let's go to the next step: querying a time series.

Reviewer 2: I'm gonna side with Vale on this one. The problem with "easy" and "easily" is: what feelings does a person feel  when they fail to do something that's been described as "easy"?

Author: Good call, this is a change for me but I will endeavor to pay attention to it and remove it from my lexicon. Thank you!

  </Collapser>
  <Collapser
    className="freq-link"
    id="time-series"
    title="4. Query time series"
  >
<SideBySide>
  <Side>

>     A common NRQL use case is querying how values change over time. This type of query provides data for line charts, area charts, and other visualizations. These "Timeseries" queries in NRDB are easy to do! We've already run some queries to explore aggregation functions `average()`, `max()`, `min()`, `sum()`, and `count()`. Now, we can take any query made with aggregation functions and add the `TIMESERIES` keyword to plot the values over time.

Reviewer 3: @jeff-colucci I think you could just remove the quotation marks around timeseries and make it lower case when you're not invoking the keyword. For example, you could probably say, "These timeseries queries...."

Author: Thank you!

```sql
SELECT average(duration) 
FROM Transaction 
SINCE 1 day ago 
TIMESERIES
```
  </Side>
  <Side>
    <img

>   title="Select the average duration since 1 day ago over with a timeseries"

Reviewer 3: @jeff-colucci looks like a typo: "over" or "with"?

  alt="A screenshot of a query that selects the average duration since 1 day ago with a timeseries"
  src={queriesnrqltutorial12}
/>
  </Side>
</SideBySide>
<SideBySide>
  <Side>

>     A "Timeseries" query breaks the data into a number of regions called buckets over the specified time period. You can have NRDB pick a default value for the width of that bucket, or you can specify your own. Let's ask NRDB to show us the average duration of application transactions, with returned data organized into 1-hour buckets.

Reviewer 3: @jeff-colucci what is the default bucket if you don't specify one?

Author: That is a good question, I will verify this with Dev or PM and add it.

>     A "Timeseries" query breaks the data into a number of regions called buckets over the specified time period. You can have NRDB pick a default value for the width of that bucket, or you can specify your own. Let's ask NRDB to show us the average duration of application transactions, with returned data organized into 1-hour buckets.

Reviewer 2: "specified" is one of those words that often slip into tech writing, but which we can easily replace with a simpler word like "chosen" or "selected"

Author: Agreed, I'll try to simplify this language impulse I have as well. Thank you!

    ```sql
    SELECT average(duration) 
    FROM Transaction 
    SINCE 1 day ago 
    TIMESERIES MAX
    ```
  </Side>
  <Side>
    <img
  title="Select the average duration since 1 day ago using the maximum number of buckets"
  alt="A screenshot of a query selecting the average duration since 1 ago using the maximum number of buckets"
  src={queriesnrqltutorial14}
/>
  </Side>
</SideBySide>
Now we're achieving new levels of visualization, breaking out from summarized numbers to line charts of data trends over time. As you can see, using `TIMESERIES` offers full control over visualizations, granularity, and averaging data over specified windows. Let's go to the next step: using the 'Where' clause.
  </Collapser>
  <Collapser
    className="freq-link"
    id="where"
    title="5. Use Where clauses"
  >
<SideBySide>
  <Side>

>     You may find that sometimes, you want your query to operate on a specific subset of NRDB data. For example, let's query our backend application's average response time. This data includes an attribute called `transactionType`, which specifies whether the transaction we timed was a "Web" or "Non-Web" (such as a cron job or queued task) transaction. Let's filter our query to only return data for the "Web" transactionType:

Reviewer 3: @jeff-colucci How do we know this is the backend or frontend?

Also, you should probably put back ticks around `transactionType`.

Author: Good point, I think I will delete backend actually since it can be applied to either one. I will add the backticks, thank you!

Author: Actually scratch that on the backticks, I think it makes more sense for us to refer to it as "transaction type" since that is actually what it is.

    ```sql
    SELECT average(duration) 
    FROM Transaction 
    WHERE transactionType='Web' 
    TIMESERIES
    ```
  </Side>
  <Side>
    <img
  title="Select the avarege duration for Web transaction types with a timeseries"
  alt="A screenshot of a query selecting the average duration for Web transaction types with a timeseries"
  src={queriesnrqltutorial15}
/>
  </Side>
</SideBySide>
<SideBySide>
  <Side>

>     You can also combine `AND` and `OR` to make more complex queries, using parentheses "(" and ")" to control how these work together.This query shows the average response time of "Web" transactions that were both faster than 100ms, and had a response code of either 200 or 302. Notice how we use parentheses to control the `OR`, which may behave differently than you'd expect.

Reviewer 3: @jeff-colucci Looks like you need a space before "This query shows..."

Also, could you explain the AND/OR scenario you thought I might expect that isn't true with NRQL?

Author: Thank you! I looked at this more and I think that the information is a little vague as it is. I have made changes.

    ```sql
    SELECT average(duration) 

>     FROM Transaction 

Reviewer 3: @jeff-colucci Are event names case sensitive in NRQL? Does it have to be "Transactions" or does "transactions" work?

Also, I don't think you need to capitalize transactions when it's used generically in a paragraph.

Author: I am 99% sure that they are not case sensitive from my own experience, but I will verify. Thank you for the capitalization callout!

Author: It looks like they are in fact case sensitive based on my research

  src={queriesnrqltutorial19}
/>
  </Side>
</SideBySide>
<SideBySide>
  <Side>
  Finally, here's a slightly more complex query that compares the quantity of Web transactions, broken down by individual applications that report to New Relic:
  ```sql
  SELECT count(*) 
  FROM Transaction 
  WHERE transactionType='Web' 
  FACET appName 
  LIMIT 5 
  SINCE 6 hours ago 
  TIMESERIES
  ```
  </Side>
  <Side>
    <img
  title="Select all web transactions from the top 5 results since 6 hours ago grouped by appName with a timeseries"
  alt="A screenshot of a query selecting all web transactions from the top 5 results since 6 hours ago grouped by appName with a timeseries"
  src={queriesnrqltutorial20}
/>
  </Side>
</SideBySide>
Grouping or "faceting" is important: it allows us to get dimensional with our data. Using the `FACET` keyword is the simplest way to do this. When applying an aggregation function, we can use `FACET` to group by any attribute, default or custom, that reports event data (just like the `WHERE` clause). We recommend tagging on custom data to your events. This allows you to more easily investigate the resulting data set.
  </Collapser>
</CollapserGroup>

> Congratulations: you've completed the first NRQL tutorial! If you've made it this far, you now know the basics of querying event data using NRQL. You can create dashboards and alerts and, if you know some React, you can even create your own Apps to run natively inside New Relic One.

Reviewer 3: @jeff-colucci you might insert a bullet list of recommended activities:

* Create dashboards and alerts.
* Explore advanced NRQL tutorials.
* Try out the NRQL course.
* Use NRQL in your own custom React apps in New Relic (By the way this seems pretty jarring for an introductory tutorial)

Author: Thank you! I do want to add guideposts for the future tutorials, but I have to create them first! Once that's done, I plan to have guideposts/recommended activities to direct the reader to the next tutorial as well as documents that explain the queries presented in the tutorials in more detail. Good callout and definitely a future plan!

Reviewer: actually, Jeff, I'd take out the 'create your own apps' stuff. basically no one does that in practice and we should de-emphasize it.

Author: Thanks, that makes it easier. I will cut out the whole final sentence. It will look a little sparse, but only until the second tutorial is out which should be one sprint at the most (I will prioritize it first so it can hopefully go out sooner).
